### PR TITLE
Feature/financial insights

### PR DESCRIPTION
**Pull request by snova-francescar**
This app demonstrates the capabilities of large language models (LLMs)
in extracting and analyzing financial data using function calling, web scraping,
and retrieval-augmented generation (RAG).

- **Stock Data Analysis**: Query and analyze stocks based on Yahoo Finance data.
- **Stock Database**: Create and query an SQL database based on Yahoo Finance data.
- **Financial News Scraping**: Scrape financial news articles from Yahoo Finance News.
- **Financial Filings Analysis**: Query and analyze financial filings based on SEC EDGAR data.
- **Generate PDF Report**: Generate a PDF report based on the saved answered queries
   or on the whole chat history.
- **Print Chat History**: Print the whole chat history.

### CODE REVIEW

File: financial_insights/src/tools.py [hunk 1]
```diff
@@ -0,0 +1,158 @@
+import datetime
```
Comment (snova-francescar) [inline financial_insights/src/tools.py:1]: @snova-codym 

This is the `src` folder.

File: financial_insights/src/function_calling.py [hunk 2]
```diff
@@ -0,0 +1,394 @@
+import json
```
Comment (snova-francescar) [inline financial_insights/src/function_calling.py:1]: @snova-codym

This is the src folder.

**Review by snova-codym — COMMENTED** [review]
Thanks and great work!  Let's continue the review in a later stage.

File: financial_insights/src/function_calling.py [hunk 3]
```diff
@@ -0,0 +1,394 @@
...
+
+        SambaVerse, SambaStudio and CoE endpoints implemented.
+
+        Returns:
+            The LLM to use.
+
+        Raises:
+            ValueError: If the LLM API is not one of `sambastudio` or `sambaverse`.
+        """
+        # SambaStudio LLM
```
[the remark this answers is not in the document]
Comment (snova-francescar) [reply financial_insights/src/function_calling.py]: I have done so in the meantime. Will push soon.

File: financial_insights/src/function_calling.py [hunk 4]
```diff
@@ -0,0 +1,394 @@
...
+from langchain_core.pydantic_v1 import BaseModel, Field
+from langchain_core.runnables import RunnableLambda
+from langchain_core.tools import StructuredTool, Tool
+
+import pydantic
+
+from financial_insights.streamlit.constants import *
+
+# Prompt template for function calling
+FUNCTION_CALLING_SYSTEM_PROMPT = """
```
[the remark this answers is not in the document]
Comment (snova-francescar) [reply financial_insights/src/function_calling.py:23]: I have created a file just for prompts now.

File: financial_insights/src/function_calling.py [hunk 5]
```diff
@@ -0,0 +1,394 @@
...
+        # Find the first JSON structure in the string
+        json_match = json_pattern.search(input_string)
+
+        # If a JSON structure was found, return it as a string
+        if json_match is not None:
+            json_str = json_match.group(1)
+            try:
+                json.loads(json_str)
+            except:
+                json_correction_prompt = """|begin_of_text|><|start_header_id|>system<|end_header_id|> You are a json format corrector tool<|eot_id|><|start_header_id|>user<|end_header_id|>
```
[the remark this answers is not in the document]
Comment (snova-francescar) [reply financial_insights/src/function_calling.py]: This has become obsolete as I have now switched to the `llama3.1` standard.

File: financial_insights/src/function_calling.py [hunk 6]
```diff
@@ -0,0 +1,394 @@
...
+                json_correction_prompt_template = PromptTemplate.from_template(json_correction_prompt)
+                json_correction_chain = json_correction_prompt_template | self.llm
+                json_str = json_correction_chain.invoke({'json': json_str})
+        else:
+            # If no JSON structure was found, return None
+            json_str = None
+            logging.warning('No tool json structure found for function calling.')
+        return json_str
+
+    def msgs_to_llama3_str(self, messages: List[BaseMessage]) -> str:
```
[the remark this answers is not in the document]
Comment (snova-francescar) [reply financial_insights/src/function_calling.py:372]: This has become obsolete as I have now switched to the llama3.1 standard.

File: financial_insights/src/tools_database.py [hunk 7]
```diff
@@ -0,0 +1,456 @@
...
+            dataframes with their corresponding purpose/type.
+
+    Returns:
+        A dictionary with company symbols as keys and a list of SQL table names as values.
+    """
+    # Connect to the SQLite database
+    engine = create_engine(f'sqlite:///{DB_PATH}')
+
+    # Create a dictionary with company names as keys and SQL tables as values
+    company_tables: Dict[str, list[str]] = dict()
```
[the remark this answers is not in the document]
Comment (snova-francescar) [reply financial_insights/src/tools_database.py:100]: Corrected and also elsewhere.

File: financial_insights/src/tools_filings.py [hunk 8]
```diff
@@ -0,0 +1,188 @@
...
+                    chunk_size=MAX_CHUNK_SIZE,
+                    chunk_overlap=64,
+                    separator=r'[.!?]',
+                    is_separator_regex=True,
+                )
+                # Split the text into chunks
+                chunks = splitter.split_text(text)
+
+                # Save chunks to csv
+                df = pandas.DataFrame(chunks, columns=['text'])
```
[the remark this answers is not in the document]
Comment (snova-francescar) [reply financial_insights/src/tools_filings.py:157]: For potential future use and retrieval.

File: financial_insights/src/tools_filings.py [hunk 9]
```diff
@@ -0,0 +1,188 @@
...
+            is not available.
+    """
+    # Checks the inputs
+    assert isinstance(user_question, str), TypeError(f'User question must be a string. Got {type(user_question)}.')
+    assert isinstance(user_request, str), TypeError(f'User question must be a string. Got {type(user_request)}.')
+    assert isinstance(ticker_symbol, str), TypeError(f'Filing type must be a string. Got {type(ticker_symbol)}.')
+    assert isinstance(filing_type, str), TypeError(f'Filing type must be a string. Got {type(filing_type)}.')
+
+    # Retrieve the filing text from SEC Edgar
+    dl = Downloader(os.environ.get('SEC_API_ORGANIZATION'), os.environ.get('SEC_API_EMAIL'))
```
[the remark this answers is not in the document]
Comment (snova-francescar) [reply financial_insights/src/tools_filings.py:82]: I will detail all the env variables in the `readme.md`.

File: financial_insights/src/tools_database.py [hunk 10]
```diff
@@ -0,0 +1,456 @@
...
+
+    Args:
+        user_query: Query to be performed on the database.
+        symbol_list: List of stock ticker symbols.
+
+    Returns:
+        The result of the query.
+    """
+    # Prompt template for the SQL queries
+    prompt = PromptTemplate.from_template(
```
[the remark this answers is not in the document]
Comment (snova-francescar) [reply financial_insights/src/tools_database.py:202]: Done.

File: financial_insights/src/tools_database.py [hunk 11]
```diff
@@ -0,0 +1,456 @@
...
+        
+        <|eot_id|><|start_header_id|>user<|end_header_id|>\
+            
+        {input}
+        <|eot_id|><|start_header_id|>assistant<|end_header_id|>"""
+    )
+
+    # Chain that receives the natural language input and the table schemas, invoke the LLM,
+    # and finally execute the SQL finder method, retrieving only the filtered SQL query
+    query_generation_chain = prompt | streamlit.session_state.fc.llm | RunnableLambda(sql_finder)
```
[the remark this answers is not in the document]
Comment (snova-francescar) [reply financial_insights/src/tools_database.py:231]: I have added a TODO, that will only be possible with larger context windows:

from langchain.chains import create_sql_query_chain
chain = create_sql_query_chain(streamlit.session_state.fc.llm, db)

https://python.langchain.com/v0.1/docs/use_cases/sql/quickstart/#convert-question-to-sql-query

[the remark this answers is not in the document]
Comment (snova-francescar) [reply financial_insights/src/tools_database.py:231]: I used Runnable here, as in the function calling kit, to retrieve the specific SQL format.

File: financial_insights/src/tools_database.py [hunk 12]
```diff
@@ -0,0 +1,456 @@
...
+
+    # Extract the names of the SQL tables that are relevant to the user query
+    selected_tables = select_database_tables(user_query, symbol_list)
+    selected_schemas = get_table_summaries_from_names(selected_tables)
+
+    # Generate the SQL query
+    query: str = query_generation_chain.invoke({'selected_schemas': selected_schemas, 'input': user_query})
+
+    # Split the SQL query into multiple queries
+    queries = query.split(';')
```
[the remark this answers is not in the document]
Comment (snova-francescar) [reply financial_insights/src/tools_database.py:241]: I added a note in the docstring.

File: financial_insights/src/tools_filings.py [hunk 13]
```diff
@@ -0,0 +1,188 @@
...
+
+    # Load the dataframe from the text file
+    try:
+        df = pandas.read_csv(CACHE_DIR + f'{filename}' + '.csv')
+    except FileNotFoundError:
+        logging.error('No scraped data found.')
+
+    # Convert DataFrame rows into Document objects
+    documents = []
+    for _, row in df.iterrows():
```
[the remark this answers is not in the document]
Comment (snova-francescar) [reply financial_insights/src/tools_filings.py:183]: I use documents only for the instantiation of the retriever, not for the LLM call.

File: financial_insights/src/tools_filings.py [hunk 14]
```diff
@@ -0,0 +1,188 @@
...
+        TypeError: If `filing_quarter` is not an integer for a `filing_type` of `10-Q`.
+        ValueError: If `filing_quarter` is not in [1, 2, 3, 4] for a `filing_type` of `10-Q`.
+        ValueError: If `filing_type` is not one of `10-K` or `10-Q`
+        ValueError: If `filing_quarter` is not provided for `10-Q` filing.
+        Exception: If a matching document for the given `filing_type`, `ticker_symbol`, and `year` is not available.
+        Exception: If a matching document for the given `filing_type`, `ticker_symbol`, `filing_quarter`, and `year`
+            is not available.
+    """
+    # Checks the inputs
+    assert isinstance(user_question, str), TypeError(f'User question must be a string. Got {type(user_question)}.')
```
[the remark this answers is not in the document]
Comment (snova-francescar) [reply financial_insights/src/tools_filings.py:76]: We might want to switch to another tool definition in the future, so for now I am leaving them.

File: financial_insights/src/tools_pdf_generation.py [hunk 15]
```diff
@@ -0,0 +1,396 @@
...
+        A list of tuples, each containing the following elements:
+            - The text.
+            - The paths of any `png` figures.
+    """
+    report_content = []
+    figure_regex = re.compile(r'financial_insights/*[^\s]+\.png')
+    directory_regex = re.compile(rf'{repo_dir}/')
+    endline_regex = re.compile(r'\n\n')
+
+    for doc in documents:
```
[the remark this answers is not in the document]
Comment (snova-francescar) [reply financial_insights/src/tools_pdf_generation.py:138]: This does a super specific thing related to my format, so I have more flexibility with this.

File: financial_insights/src/utilities_retrieval.py [hunk 16]
```diff
@@ -0,0 +1,73 @@
...
+        TypeError: If user request is not string or documents are not list of strings.
+    """
+    assert isinstance(user_request, str), TypeError('user_request must be a string.')
+    assert isinstance(documents, list), TypeError(f'documents must be a list of strings. Got {type(documents)}.')
+    assert all(isinstance(doc, Document) for doc in documents), TypeError(
+        f'All documents must be of type `langchain.schema.Document`.'
+    )
+
+    # Instantiate the embedding model
+    embedding_model = SentenceTransformerEmbeddings(model_name='paraphrase-mpnet-base-v2')
```
[the remark this answers is not in the document]
Comment (snova-francescar) [reply financial_insights/src/utilities_retrieval.py:39]: Added a TODO.

File: financial_insights/prompts/llama30b-web_crawling_data_retriever.yaml
Comment (snova-codym) [inline financial_insights/prompts/llama30b-web_crawling_data_retriever.yaml:1]: Is Llama 2 being used here?

**Review by snova-codym — COMMENTED** [review]
Good from here.

File: financial_insights/prompts/llama30b-web_crawling_data_retriever.yaml
Comment (snova-francescar) [reply financial_insights/prompts/llama30b-web_crawling_data_retriever.yaml:1]: Good spot here! I'll upgrade it.

Comment (snova-francescar) [reply financial_insights/prompts/llama30b-web_crawling_data_retriever.yaml:1]: Actually I have removed that prompt and the deprecated `RetrieveQA`. I am now using `from langchain.chains import create_retrieval_chain`. See [migration guide](https://python.langchain.com/v0.2/docs/versions/migrating_chains/retrieval_qa/).

File: financial_insights/src/tools.py [hunk 17]
```diff
@@ -0,0 +1,512 @@
...
+        general_logger.setLevel(logging.INFO)
+        console_handler = logging.StreamHandler()
+        console_handler.setLevel(logging.INFO)
+        general_formatter = logging.Formatter('%(asctime)s - %(message)s')
+        console_handler.setFormatter(general_formatter)
+        general_logger.addHandler(console_handler)
+    return general_logger
+
+
+logger = get_general_logger()
```
Comment (snova-codym) [inline financial_insights/src/tools.py:56]: Remove?

File: financial_insights/streamlit/readme.md [hunk 18]
```diff
@@ -0,0 +1,124 @@
+
```
Comment (snova-codym) [inline financial_insights/streamlit/readme.md:1]: Maybe this readme file should exist in the higher level directory for financial assistant?

File: financial_insights/streamlit/readme.md [hunk 19]
```diff
@@ -0,0 +1,124 @@
...
+Clone the starter kit repo.
+```
+git clone https://github.com/sambanova/ai-starter-kit.git
+```
+
+## Set up the inference endpoint, configs and environment variables
+
+The next step is to set up your environment variables to use one of the models available from SambaNova. If you're a current SambaNova customer, you can deploy your models with SambaStudio. If you are not a SambaNova customer, you can self-service provision API endpoints using SambaNova Fast API or Sambaverse. Note that Sambaverse, although freely available to the public, is rate limited and will not have fast RDU optimized inference speeds.
+
+- If using **SambaNova Fast-API** Please follow the instructions [here](../README.md#use-sambanova-fast-api-option-1) for setting up your environment variables.
```
Comment (snova-codym) [inline financial_insights/streamlit/readme.md:61]: Not sure that this and the following links are working correctly.

File: financial_insights/streamlit/utilities_app.py [hunk 20]
```diff
@@ -0,0 +1,495 @@
...
+
+def save_output_callback(
+    response: Union[str, List[str], Dict[str, str]],
+    save_path: str,
+    user_request: Optional[str] = None,
+) -> None:
+    """Save the output callback for streamlit button."""
+    # Check the inputs
+    assert isinstance(response, (str, list, dict, tuple, pandas.DataFrame)), TypeError(
+        f'Response must be a string, a list, a dictionary, or a tuple. Got type {type(response)}'
```
Comment (snova-codym) [inline financial_insights/streamlit/utilities_app.py:97]: What about dataframe in this message?

**Review by snova-codym — COMMENTED** [review]
Only very minor commentary!  Great work.  Was your readme intended to be in the streamlit folder?  I think it may have broken some links.

File: financial_insights/streamlit/utilities_app.py (hunk 20 above)
Comment (snova-francescar) [reply financial_insights/streamlit/utilities_app.py:97]: Added, and also added series.

File: financial_insights/streamlit/readme.md (hunk 18 above)
Comment (snova-francescar) [reply financial_insights/streamlit/readme.md:1]: Yes, thank you for noticing this.

**Review by snova-kwasia — APPROVED** [review]
Thank you @snova-francescar for the hard work and @snova-codym for the review!

**Pull request merged**